### PR TITLE
Remove base_param_map

### LINKED ISSUE #587: substitute_fn argument in the substitute effect handler.
Is the way to use substitute_fn expected? 

I was expecting that values returned by substitute_fn will be unchanged. However, from the code it appears that these values will be base_value and will be transformed. This is not mentioned in the documentation.

https://github.com/pyro-ppl/numpyro/blob/25d055ad4b5a5e85a02159af87e8db1451243217/numpyro/handlers.py#L432

### LINKED ISSUE #594: AutoContinuous not working correctly with TransformedDistribution
The implementation of `AutoContinuous` will not work with `TransformedDistribution` objects, because the transformations specified for the distribution seem to be never applied. We encountered this issue with a `LogNormal` distribution and here is a minimal example:
```python
import jax.numpy as np
import jax

import numpyro.distributions as dist
from numpyro.primitives import sample
from numpyro.infer import SVI, ELBO
from numpyro.contrib.autoguide import AutoDiagonalNormal
from numpyro.optim import SGD

import numpy as onp

data = .3 * onp.random.randn(100) + 1.

def model(x):
    mu = sample('mus', dist.Normal(0., 1.))
    sig = sample('sig', dist.LogNormal(0., 1.))
    sample('x', dist.Normal(mu, sig), obs=x)

guide = AutoDiagonalNormal(model)

optim = SGD(1e-4)
svi = SVI(model, guide, optim, ELBO())

rng_key = jax.random.PRNGKey(0) # works with 1, does not work with 0
svi_state = svi.init(rng_key, data)
init_svi_state = svi_state

for _ in range(10):
    svi_state, loss = svi.update(svi_state, data)
    print(loss)

```

If the key is seeded with 0, this will produce `nan` values (but it will work fine
for a seed of 1, or even if the order of mu and sig in the model is changed for
seed 0). The reason is that for a seed of 0 the guide samples negative initial values 
for sigmas (to which `LogNormal.log_prob` then applies a log).

We believe the issue lies in `AutoContinuous._setup_prototype` (particularly [line 154ff](https://github.com/pyro-ppl/numpyro/blob/5755a2ea845eb7e1e7878334cb711f70317b2512/numpyro/contrib/autoguide/__init__.py#L152))
where, for sample sites with intermediates, the guide just extracts a transform
from unconstrained optimization space to the support of a base distribution and
stores it in its `_inv_transforms` lookup, ignoring all other transformation the
site's distribution applies. The `__call__` implementation therefore applies
only the transform from unconstrained space to the constraint support of
the base distribution, not the target distribution.

The `LogNormal` is implemented as `TransformedDistribution(Normal, ExpTransform())`,
so it's base distribution has real support and therefore the guide extracts an
`IdentityTransform`. The `ExpTransform` seems to be never applied (at least we
could not pinpoint a location where that would be the case), thus the value
returned by the guide can be negative.

Note: `Pareto` is an interesting case, because it doesn't fall into the same
trap. It is a `TransformedDistribution`, applying an `ExpTransform` and an `AffineTransform`,
which the guide ignores, but it's base distribution is `Exponential`, which
has positive support and thus the guide will extract an `ExpTransform` anyways,
ensuring that the values it samples are positive (albeit probably not scaled properly).

~~A pull request with a suggested bugfix will follow soon.~~
Edit: We're not sure what would be the best fix for this.

From how the `__call__` function is written, it seems that a straightforward fix would be to change [line 153](https://github.com/pyro-ppl/numpyro/blob/5755a2ea845eb7e1e7878334cb711f70317b2512/numpyro/contrib/autoguide/__init__.py#L153) from `transform = biject_to(site['fn'].base_dist.support)`
to
`transform = ComposeTransform([biject_to(site['fn'].base_dist.support)] + site['fn'].transforms`, however, this comes with several caveats:
- it only works for `TransformedDistribution` classes that define a `transforms` attribute and will crash for all other distributions that sample intermediates
- there seems to be some way of applying the correct transformations to the latent sample from optimization space, because methods returning parameter values from an `SVIState` do that already. `_unpack_and_constrain` seems to be doing that maybe, but we're not quite sure and wouldn't know how to integrate it into `__call__`, which is currently not using `_unpack_and_constrain` but doing it's own stuff.

Edit 2: We have not much experience with Pyro, but from a quick look it seems that the implementation there is almost identical (minus the obvious structural differences and some case distinctions), so it would be interesting to know whether the same problems persists there, but we currently lack the time to do so..

### PR DESCRIPTION
**Pull request by fehiepsi**
With #599, we no longer need to do "automatic reparameterization" behind the scene for dynamic support. This PR removes stuff related to base_param_map.

Fixes #587 and fixes #594.

### TODO
- [x] remove stuffs in `infer.util`
- [x] make sure all tests pass, probably with the usage of "reparam"
- [x] remove dynamic support in autoguide

### CODE REVIEW

File: numpyro/contrib/autoguide/__init__.py [hunk 1]
```diff
@@ -203,44 +185,16 @@ def __call__(self, *args, **kwargs):
...
-        if sample_shape:
-            latent_sample = np.reshape(latent_sample, (-1, np.shape(latent_sample)[-1]))
-            unpacked_samples = vmap(unpack_single_latent)(latent_sample)
-            unpacked_samples = tree_map(lambda x: np.reshape(x, sample_shape + np.shape(x)[1:]),
-                                        unpacked_samples)
-        else:
-            unpacked_samples = unpack_single_latent(latent_sample)
-        return unpacked_samples
+    def _unpack_and_constrain(self, latent_sample):
+        return transform_fn(self._inv_transforms, self._unpack_latent(latent_sample))
```
Comment (fehiepsi) [inline numpyro/contrib/autoguide/__init__.py:197]: We save a bunch of code here. :D

File: numpyro/contrib/autoguide/__init__.py [hunk 2]
```diff
@@ -610,42 +551,3 @@ def _get_transform(self):
             arnn = numpyro.module('{}_arn__{}'.format(self.prefix, i), arn, (self.latent_size,))
             flows.append(BlockNeuralAutoregressiveTransform(arnn))
         return ComposeTransform(flows)
-
-
-class AutoContinuousELBO(ELBO):
```
Comment (fehiepsi) [inline numpyro/contrib/autoguide/__init__.py:615]: We no longer need this class. cc @lumip

File: numpyro/contrib/reparam.py [hunk 3]
```diff
@@ -55,6 +55,14 @@ def process_message(self, msg):
         new_fn, value = reparam(msg["name"], msg["fn"], msg["value"])
 
         if value is not None:
+            if new_fn is None:
+                msg['type'] = 'deterministic'
+                msg['value'] = value
+                for key in list(msg.keys()):
+                    if key not in ('type', 'name', 'value'):
+                        del msg[key]
+                return
```
Comment (fehiepsi) [inline numpyro/contrib/reparam.py:64]: Returns `deterministic` site here instead of `sample` site. This way, MCMC can report results for deterministic nodes.

File: numpyro/distributions/transforms.py [hunk 4]
```diff
@@ -498,7 +498,7 @@ def __call__(self, x):
         batch_shape = x.shape[:-1]
         if batch_shape:
             unpacked = vmap(self.unpack_fn)(x.reshape((-1,) + x.shape[-1:]))
-            return tree_map(lambda x: np.reshape(x, batch_shape + x.shape[-1:]), unpacked)
+            return tree_map(lambda z: np.reshape(z, batch_shape + z.shape[1:]), unpacked)
```
Comment (fehiepsi) [inline numpyro/distributions/transforms.py:501]: Fix a `reshape` bug (`unpacked`'s event shape starts at 1, rather than -1).

File: numpyro/handlers.py [hunk 5]
```diff
@@ -259,25 +257,28 @@ class condition(Messenger):
        >>> assert exec_trace['a']['value'] == -1
        >>> assert exec_trace['a']['is_observed']
     """
-    def __init__(self, fn=None, param_map=None, substitute_fn=None):
-        self.substitute_fn = substitute_fn
+    def __init__(self, fn=None, param_map=None, condition_fn=None):
```
Comment (fehiepsi) [inline numpyro/handlers.py:260]: I think this is typo. In the docs, this arg is `condition_fn`.

File: numpyro/infer/hmc_util.py [hunk 6]
```diff
@@ -209,8 +209,8 @@ def update_fn(step_size, inverse_mass_matrix, state):
     return init_fn, update_fn
 
 
-def find_reasonable_step_size(potential_fn, kinetic_fn, momentum_generator, inverse_mass_matrix,
-                              position, rng_key, init_step_size):
+def find_reasonable_step_size(potential_fn, kinetic_fn, momentum_generator,
+                              init_step_size, inverse_mass_matrix, position, rng_key):
```
Comment (fehiepsi) [inline numpyro/infer/hmc_util.py:213]: I just reorder the position of `init_step_size` to clean up the code (using the utility `identity`).

File: numpyro/infer/initialization.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from functools import partial
```
Comment (fehiepsi) [inline numpyro/infer/initialization.py:4]: Just copy 1-1 from `infer/util.py` file for clarity.

File: numpyro/contrib/reparam.py [hunk 8]
```diff
@@ -55,6 +55,14 @@ def process_message(self, msg):
         new_fn, value = reparam(msg["name"], msg["fn"], msg["value"])
 
         if value is not None:
+            if new_fn is None:
+                msg['type'] = 'deterministic'
+                msg['value'] = value
+                for key in list(msg.keys()):
+                    if key not in ('type', 'name', 'value'):
```
Comment (neerajprad) [inline numpyro/contrib/reparam.py:62]: Could you explain why we need to delete here?

File: numpyro/contrib/reparam.py (hunk 3 above)
Comment (neerajprad) [reply numpyro/contrib/reparam.py:64]: How do we handle this in Pyro?

File: numpyro/contrib/reparam.py (hunk 8 above)
Comment (fehiepsi) [reply numpyro/contrib/reparam.py:62]: We don't need to delete, but `deterministic` default msg doesn't hold those keys, so I delete for consistency.

File: numpyro/contrib/reparam.py (hunk 3 above)
Comment (fehiepsi) [reply numpyro/contrib/reparam.py:64]: In Pyro, we use Delta observation `sample` site because Pyro doesn't have `deterministic` site yet.

Comment (neerajprad) [reply numpyro/contrib/reparam.py:64]: I see. So with this change, we'll need to wrap this in `Predictive` to get the values at deterministic nodes still, right?

Comment (fehiepsi) [reply numpyro/contrib/reparam.py:64]: For both Delta obs or deterministic, we need to use Predictive. But with deterministic, we can retrace the model at postprocess_fn, as we currently do, so no need to use Predictive. I think we need to do the same for autoguide but I'll address it separately.

File: numpyro/contrib/autoguide/__init__.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/contrib/autoguide/__init__.py:197]: @neerajprad I'll add a commit to record `deterministic` sites for autoguide. I think we did it previously here.

**Comment by fehiepsi** [comment]
@neerajprad I have made a change at `AutoGuide._unpack_and_constrain` method to return deterministic sites for `sample_posteriors`, `median`, `quantiles` methods. This makes autoguide consistent with mcmc. WDYT?

**Comment by neerajprad** [comment]
> @neerajprad I have made a change at AutoGuide._unpack_and_constrain method to return deterministic sites for sample_posteriors, median, quantiles methods. This makes autoguide consistent with mcmc. WDYT?

In principle, this is a fine change but I would like to discuss what we get by doing this and alternatives, if any. Some of the tests are now failing, btw. How about we make this change and discuss in a separate PR instead (just the autoguide related change; MCMC looks fine)?

**Comment by fehiepsi** [comment]
Yes, let's discuss it when we move autoguide to infer or at #609 (the autoguide part).

**Comment by neerajprad** [comment]
Sounds good. LGTM once the autoguide changes are reverted.
